### PR TITLE
planner/cascades: add transformation rule ProjectionElimination

### PR DESCRIPTION
**Pull request by t0350**
### What problem does this PR solve? 

add transformation rule for planner/cascades, try to eliminate the projection whose output is the same with its child.

### What is changed and how it works?

to convert `Projection -> Any` to `Any` when it is possible, will promote the child group's expression.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by t0350** [comment]
Not quite sure whether I understand the transformation correctly. Thanks for any comment.

**Comment by francis0407** [comment]
> Not quite sure whether I understand the transformation correctly. Thanks for any comment.

Yes, this is definitely what we want. Thanks for your contribution. I will review this PR later.

File: planner/cascades/transformation_rules.go [hunk 1]
```diff
@@ -591,3 +594,43 @@ func (r *PushSelDownJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.Grou
 	newJoinExpr.SetChildren(leftGroup, rightGroup)
 	return []*memo.GroupExpr{newJoinExpr}, true, false, nil
 }
+
+// SameProjectionElimination eliminate the projection.
+type SameProjectionElimination struct {
```
Comment (winoros) [inline planner/cascades/transformation_rules.go]: Simplify the name to `ProjectionElimination`?

Comment (francis0407) [reply planner/cascades/transformation_rules.go]: Or `EliminateProjection`? We have been using names like `PushXXXDown`, `EnumeratePath`.

Comment (t0350) [reply planner/cascades/transformation_rules.go]: OK, I will use `EliminateProjection`.

File: planner/cascades/transformation_rules.go [hunk 2]
```diff
@@ -591,3 +594,43 @@ func (r *PushSelDownJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.Grou
 	newJoinExpr.SetChildren(leftGroup, rightGroup)
 	return []*memo.GroupExpr{newJoinExpr}, true, false, nil
 }
+
+// SameProjectionElimination eliminate the projection.
```
[the remark this answers is not in the document]
Comment (t0350) [reply planner/cascades/transformation_rules.go]: Fixed.

File: planner/cascades/transformation_rules_test.go [hunk 3]
```diff
@@ -157,3 +157,24 @@ func (s *testTransformationRuleSuite) TestTopNRules(c *C) {
...
 	testGroupToString(input, output, s, c)
 }
+
+func (s *testTransformationRuleSuite) TestProjectionElimination(c *C) {
+	s.optimizer.ResetTransformationRules(map[memo.Operand][]Transformation{
+		memo.OperandProjection: {
+			NewRuleEliminateProjection(),
+		},
+		memo.OperandDataSource: {
+			NewRuleEnumeratePaths(),
```
Comment (francis0407) [inline planner/cascades/transformation_rules_test.go]: I think this is unnecessary. We only need `NewRuleEliminateProjection` here?

Comment (francis0407) [reply planner/cascades/transformation_rules_test.go]: But it is ok to be here. Don't need to change it then.

Comment (t0350) [reply planner/cascades/transformation_rules_test.go]: OK, I just copied from the upper test function. Fixed. The test output simplified a lot.

**Pull request merged**